frontend/CODE_GUIDELINES.md

```diff
@@ -0,0 +1,105 @@
+  - It's easier to abstract code than to remove an abstraction.
+- Break switch statements into their own functions with returns rather than `break;`.
+
+## File structure
+
+- Higher-order functions go at the top of the file when possible.
+- Named exports are preferred to default exports. It's easier to control naming.
+- Inline exports are preferred to the define-then-export pattern.
+
+## Testing
+
+Flakey tests are the worst. The following patterns ensure that tests don't interact.
+
+- Prefer many `describe` blocks.
+- Setup and teardown belong in `beforeEach`/`afterEach` blocks if possible. 
+- `beforeAll`/`afterAll` can be a reasonable fallback.
+- Some setup may be required in `it` blocks, but try to minimize it
+- Prefer `uuid` strings to `testing-testing-123`-style strings.
+- Don't over-test the DOM structure. 
+  - Focus on testing functionality.
+  - Simple components need tests too, mostly to make sure we don't break their imports.
+  - Testing is as much about easy refactors as anything. We need to be able to refactor confidently.
+- Don't be afraid to instrument your code for testing purposes.
+- Get good at mocking imports.
+  - Declare your mocks variables at the top of the file with the `let mock*` pattern.
+  - `jest.mock()` calls should follow the `let mock*` declarations.
+  - Jest automatically hoists your `jest.mock()` calls to the top of the file, so avoid confusion by keeping them near the top of the file. 
+  - Avoid `mockFn.mockImplementation` and `mockFn.mockImplementationOnce`. They're not **bad**, but they're more verbose and a bit harder to read than the `let mock*` pattern.
+
+## Javascript
+
+- Prefer `const` over `let`. Avoid `var`.
+- Prefer named functions, i.e., `function useSomething(){}` over anonymous functions.
+- Arrow functions are best for one-liners.
+- Write functional JS. `class` and `this` are unnecessary in most modern JavaScript.
+- Avoid mutating objects. This is bug-avoidance-rule-number-one in JavaScript.
+  - Object mutation is fine if it's visually close to the object declaration.
+  - There are times when object mutation simplifies the code, but BE CAREFUL AND EXPLICIT.
+- Be wary of using closure-scope for variable access. It's not functional and couples functions together, making refactors difficult.
+- Prefer "dot-syntax" over excessive destructuring.
+  - Well-crafted destructures can be very readable.
+  - Too much destructuring can kill your readability.
+  - For instance, always use `import constants from '~/constants';` This enables confident find/replace changes to the constants, because you can search for `constants.SOME_KEY`.
+  - Destructuring removes the context of the variable. The reader no longer knows where it came from. This is fine in vertically-adjacent code... but it can be hard to read otherwise.
+
+## Functions
+
+- Prefer a single return. There are definitely exceptions, but early returns can make debugging difficult.
```

Comment:
Not sure I agree here, but would maybe a test case could convince me. Think early returns can make it easier on readability and debugging